libPhoneNumber-GeocodingParser/libPhoneNumber-GeocodingParser/NBSQLiteDatabaseConnection.m

```diff
@@ -0,0 +1,111 @@
+    @"CREATE TABLE IF NOT EXISTS geocodingPairs%@ (ID INTEGER PRIMARY KEY "
+    @"AUTOINCREMENT, "
+    @"NATIONALNUMBER TEXT, DESCRIPTION TEXT)";
+static NSString *const createIndexStatement = @"CREATE INDEX IF NOT EXISTS nationalNumberIndex ON "
+                                              @"geocodingPairs%@(NATIONALNUMBER)";
+
+- (instancetype)initWithCountryCode:(NSString *)countryCode
+                       withLanguage:(NSString *)language
+                withDestinationPath:(NSURL *)destinationPath {
+  self = [super init];
+  if (self != nil) {
+    NSString *databasePath = [[NSString alloc]
+        initWithString:[NSString stringWithFormat:@"%@/%@.db", destinationPath.path, language]];
+    _sqliteDatabaseCode = sqlite3_open([databasePath UTF8String], &_DB);
+
+    if (_sqliteDatabaseCode == SQLITE_OK) {
+      [self createTable:countryCode];
+      const char *formattedPreparedStatement =
+          [[NSString stringWithFormat:insertPreparedStatement, countryCode] UTF8String];
+      sqlite3_prepare_v2(_DB, formattedPreparedStatement, -1, &_insertStatement, NULL);
+    } else {
+      NSLog(@"Cannot open database at desired location: %@. \n"
+            @"SQLite3 Error Message: %s",
+            destinationPath, sqlite3_errstr(_sqliteDatabaseCode));
+    }
+  }
+  return self;
+}
+
+- (void)addEntryToDB:(NSString *)phoneNumber
+     withDescription:(NSString *)description
+     withCountryCode:(NSString *)countryCode {
+  if (_sqliteDatabaseCode != SQLITE_OK) {
+    NSLog(@"Cannot open database. Failed to add entry. \n"
+          @"SQLite3 Error Message: %s",
+          sqlite3_errstr(_sqliteDatabaseCode));
+    return;
+  }
+  int commandResults = [self createInsertStatement:phoneNumber withDescription:description];
+  if (commandResults != SQLITE_OK) {
+    NSLog(@"Error when creating insert statement: %s", sqlite3_errstr(commandResults));
+  }
+  sqlite3_step(_insertStatement);
+}
+
+- (void)dealloc {
+  sqlite3_finalize(_insertStatement);
+  sqlite3_close_v2(_DB);
+}
+
+- (void)createTable:(NSString *)countryCode {
+  NSString *createTable = [NSString stringWithFormat:createTablePreparedStatement, countryCode];
+
+  const char *sqliteCreateTableStatement = [createTable UTF8String];
+  char *sqliteErrorMessage;
+  if (sqlite3_exec(_DB, sqliteCreateTableStatement, NULL, NULL, &sqliteErrorMessage) != SQLITE_OK) {
+    NSLog(@"Error creating table, %s", sqliteErrorMessage);
+    return;
+  }
+
+  NSString *createIndexQuery = [NSString stringWithFormat:createIndexStatement, countryCode];
+  const char *sqlCreateIndexStatement = [createIndexQuery UTF8String];
+  if (sqlite3_exec(_DB, sqlCreateIndexStatement, NULL, NULL, &sqliteErrorMessage) != SQLITE_OK) {
+    NSLog(@"Error occurred when applying index to nationalnumber column: %s", sqliteErrorMessage);
+    return;
+  }
+}
+
+- (int)resetInsertStatement {
+  return sqlite3_reset(_insertStatement) | sqlite3_clear_bindings(_insertStatement);
```

Comment:
Not sure if this is a good idea though. since if it fails, we can't really know what the error code actually is. Maybe just do 2 ifs here:
```
if reset != ok {
  return;
}
if clear_bindings != ok {
  return;
}
```